app/models/service_provider.rb

```diff
@@ -43,4 +45,20 @@ def encryption_opts
   def live?
     active? && approved?
   end
+
+  private
+
+  def redirect_uris_are_parsable
+    return if redirect_uris.blank?
+
+    redirect_uris.each do |uri|
+      begin
+        next if uri =~ URI::DEFAULT_PARSER.regexp[:ABS_URI] && URI.parse(uri)
+        raise URI::InvalidURIError
+      rescue URI::BadURIError, URI::InvalidURIError
+        errors.add(:redirect_uris, :invalid)
+        break
+      end
+    end
```

Comment:
If the URI contains spaces, it will fail (as expected), but we can easily fix that before we parse the URI. From a UX perspective, I think it's better for us to fix formatting issues for the user. This would mean a `before_validation` callback that calls `.strip` on the URI. I'm generally not crazy about callbacks, but I think it's fine in this situation since it only operates on the ServiceProvider object. Thoughts?

Reply:
The user here is the dashboard app which should not be sending us malformed URLs. I agree that we should be removing spaces from the URLs for the user, but we should be doing that on the dashboard before we write them into the db, and not in the IDP.

Reply:
If there is validation/enforcement we can move to the dashboard instead to guarantee structure of the API output, happy to do that so we don't have to build some elaborate guards in the IdP.